### PR TITLE
Add ability to Eviction to resize Eden/Main ratio

### PR DESCRIPTION
**Pull request by blsemo**
PR to initially discuss the approach to change the size of the window cache.

### CODE REVIEW

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java [hunk 1]
```diff
@@ -441,11 +441,21 @@ protected void lazySetMainProtectedMaximum(long maximum) {
    * to eagerly evicts entries until the cache shrinks to the appropriate size.
    */
   @GuardedBy("evictionLock")
-  void setMaximum(long maximum) {
+  void setMaximum(long maximum){
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: whitespace

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java [hunk 2]
```diff
@@ -3156,7 +3166,8 @@ Object writeReplace() {
       @Override public long getMaximum() {
         return cache.maximum();
       }
-      @Override public void setMaximum(long maximum) {
+      @Override public void setMaximum(long maximum){ setMaximum(maximum, PERCENT_MAIN);}
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: whitespace.  However, I suspect their coding style is to put onto distinct lines rather than all on one?

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java [hunk 3]
```diff
@@ -441,11 +441,21 @@ protected void lazySetMainProtectedMaximum(long maximum) {
    * to eagerly evicts entries until the cache shrinks to the appropriate size.
    */
   @GuardedBy("evictionLock")
-  void setMaximum(long maximum) {
+  void setMaximum(long maximum){
+      setMaximum(maximum, PERCENT_MAIN);
+  }
+
+  /**
+   * Sets the maximum weighted size of the cache and the ration between eden and LFU.
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: ration -> ratio

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java [hunk 4]
```diff
@@ -167,6 +167,25 @@ default OptionalInt weightOf(@NonNull K key) {
...
+     * the appropriate size.
+     * <p>
+     * This version also sets the ratio between LRU eden and the LFU cache
+     * <p>
+     * Note that some implementations may have an internal inherent bound on the maximum total size.
+     * If the value specified exceeds that bound, then the value is set to the internal maximum.
+     *
+     * @param maximum the maximum, interpreted as weighted or unweighted size depending on how this
+     *        cache was constructed
+     * @param percentMain relative size of the main (LFU) cache as percentage (0 &lt; percentMain &lt; 1)
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java]: Less than or equal to in both cases I think

File: caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+public class TestEdenResize {
+    @Test
+    public void TestResizingEden(){
+        Cache<Integer, Integer> cache = Caffeine.newBuilder().maximumSize(1024).build();
+        cache.put(1,1);
+        Policy.Eviction<Integer, Integer> eviction = cache.policy().eviction().orElse(null);
+        Assert.assertNotNull(eviction);
+        Assert.assertNotNull(cache.getIfPresent(1));
+        if (eviction != null){
```
Comment (rick-r3) [inline caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java]: Whitespace.  Maybe just reformat your changed lines.

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java (hunk 2 above)
Comment (blsemo) [reply caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: Hah! You are right and I got tricked by IntelliJ folding!

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java [hunk 6]
```diff
@@ -3156,7 +3166,8 @@ Object writeReplace() {
       @Override public long getMaximum() {
         return cache.maximum();
       }
-      @Override public void setMaximum(long maximum) {
+      @Override public void setMaximum(long maximum){ setMaximum(maximum, PERCENT_MAIN);}
+      @Override public void setMaximum(long maximum, double percentMain) {
         cache.evictionLock.lock();
         try {
           cache.setMaximum(maximum);
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: Should this be setting the `percentMain` too ?  Anyway we can invent a test, with a really small cache, to check?

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java (hunk 3 above)
Comment (rick-r3) [reply caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: Maybe the explanation should be less hand-wavy and explicitly say it defines the percentage given over to the LFU.

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java [hunk 7]
```diff
@@ -441,11 +441,21 @@ protected void lazySetMainProtectedMaximum(long maximum) {
...
+  }
+
+  /**
+   * Sets the maximum weighted size of the cache and the ration between eden and LFU.
+   *  The caller may need to perform a maintenance cycle
+   * to eagerly evicts entries until the cache shrinks to the appropriate size.
+   */
+  @GuardedBy("evictionLock")
+  void setMaximum(long maximum, double mainPercentage) {
     requireArgument(maximum >= 0);
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: require bounds on the percentage 0 <= mainPercentage <= 1.0d

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java [hunk 8]
```diff
@@ -167,6 +167,25 @@ default OptionalInt weightOf(@NonNull K key) {
...
+     * @throws IllegalArgumentException if the maximum size specified is negative
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java]: Also (when you make it) thrown if the percentMain is out of range

File: caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+            eviction.setMaximum(eviction.getMaximum(), .5d);
+        }
+        Assert.assertNotNull(cache.getIfPresent(1));
```
Comment (rick-r3) [inline caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java]: Add a test with negative percent and 1.1d ?

File: caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        boolean missing = false;
+        for (int counter = 1000; counter < 1500; ++counter) {
+            if (!cm.keySet().contains(counter)) {
+                missing = true;
+            }
+        }
+        Assert.assertTrue("All keys made it to the main cache!", missing);
+    }
+
+
```
Comment (rick-r3) [inline caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java]: remove blank lines

File: caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    @Test
+    public void Test50PercentWindow() throws InterruptedException {
+        Cache<Integer, Integer> cache = Caffeine.newBuilder().maximumSize(1000).build();
+        Policy.Eviction<Integer, Integer> eviction = cache.policy().eviction().get();
+        eviction.setMaximum(eviction.getMaximum(), .5d);
+        for (int counter = 0; counter < 2000; ++counter) {
+            cache.put(counter, counter);
+        }
+
+
```
Comment (rick-r3) [inline caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java]: Remove blank lines

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java [hunk 12]
```diff
@@ -167,6 +167,26 @@ default OptionalInt weightOf(@NonNull K key) {
...
+     * If the value specified exceeds that bound, then the value is set to the internal maximum.
+     *
+     * @param maximum the maximum, interpreted as weighted or unweighted size depending on how this
+     *        cache was constructed
+     * @param percentMain relative size of the main (LFU) cache as percentage (0 &le; percentMain &le; 1)
+     * @throws IllegalArgumentException if the maximum size specified is negative or percentMain is less
+     *        than 0 or greater than 1
+     */
+    void setMaximum(@NonNegative long maximum, double percentMain);
+
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java:188]: Remove blank line

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java [hunk 13]
```diff
@@ -3156,10 +3168,14 @@ Object writeReplace() {
       @Override public long getMaximum() {
         return cache.maximum();
       }
-      @Override public void setMaximum(long maximum) {
+      @Override public void setMaximum(long maximum)
+      {
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java]: Put the opening brace on line above (seems to be their style)

File: caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java [hunk 14]
```diff
@@ -167,6 +167,25 @@ default OptionalInt weightOf(@NonNull K key) {
...
+     * the appropriate size.
+     * <p>
+     * This version also sets the ratio between LRU eden and the LFU cache
+     * <p>
+     * Note that some implementations may have an internal inherent bound on the maximum total size.
+     * If the value specified exceeds that bound, then the value is set to the internal maximum.
+     *
+     * @param maximum the maximum, interpreted as weighted or unweighted size depending on how this
+     *        cache was constructed
+     * @param percentMain relative size of the main (LFU) cache as percentage (0 &le; percentMain &le; 1)
```
Comment (rick-r3) [inline caffeine/src/main/java/com/github/benmanes/caffeine/cache/Policy.java]: "as percentage of maximum" perhaps?

File: caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        Policy.Eviction<Integer, Integer> eviction = cache.policy().eviction().get();
+        eviction.setMaximum(eviction.getMaximum(), 0d);
+        for (int counter = 0; counter < 2000; ++counter) {
+            cache.put(counter, counter);
+        }
+
+        // need to wait for a moment to give the cache time to evict stuff.
+        sleep(1000);
+
+        // if the window is 100% of the cache, the last 1000 entries should be in the cache. There can be a  temporary
```
Comment (rick-r3) [inline caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java]: double space before "temporary"

File: caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    public void Test100PercentWindow() throws InterruptedException {
+        Cache<Integer, Integer> cache = Caffeine.newBuilder().maximumSize(1000).build();
+        Policy.Eviction<Integer, Integer> eviction = cache.policy().eviction().get();
+        eviction.setMaximum(eviction.getMaximum(), 0d);
+        for (int counter = 0; counter < 2000; ++counter) {
+            cache.put(counter, counter);
+        }
+
+        // need to wait for a moment to give the cache time to evict stuff.
+        sleep(1000);
```
Comment (rick-r3) [inline caffeine/src/test/java/com/github/benmanes/caffeine/cache/TestEdenResize.java]: Wondering if the executor allows us to force this to happen rather than racy sleep.

**Comment by blsemo** [comment]
Closing and raising new PR against the real master branch.
